app/uk/gov/hmrc/eusubsidycompliancestub/controllers/UndertakingController.scala

```diff
@@ -90,145 +94,102 @@ class UndertakingController @Inject() (
+        Ok(
+          Json.toJson(
+            CreateUndertakingApiResponse("102", s"Invalid EORI number $eori")
           )
-        )
-        Future.successful(Ok(Json.toJson(invalidEori)))
+        ).toFuture
 
       case e if e.endsWith("555") =>
-        val missingPostcode: JsValue = Json.obj(
-          "createUndertakingResponse" -> Json.obj(
-            "responseCommon" -> badResponseCommon(
-              "113",
-              s"Postcode missing for the address"
-            )
-          )
-        )
-        Future.successful(Ok(Json.toJson(missingPostcode)))
+        Ok(Json.toJson(CreateUndertakingApiResponse("113", s"Postcode missing for the address"))).toFuture
+
       //create an Undertaking with lastSubsidyUsageUpdt which is 77 days older than today i.e between the range of 76-90 days
       case f if f.endsWith("444") =>
         val JsSuccess(undertaking, _) = Json.fromJson(json)(undertakingRequestReads)
         val madeUndertaking = EisService.makeUndertaking(undertaking, eori, LocalDate.now.minusDays(77).some)
         Store.undertakings.put(madeUndertaking)
-        Future.successful(Ok(Json.toJson(madeUndertaking.reference.get)(eisCreateUndertakingResponse)))
+        Ok(Json.toJson(CreateUndertakingApiResponse(madeUndertaking.reference.get))).toFuture
 
       case _ =>
         val JsSuccess(undertaking, _) = Json.fromJson(json)(undertakingRequestReads)
         val madeUndertaking = EisService.makeUndertaking(undertaking, eori)
         Store.undertakings.put(madeUndertaking)
-        Future.successful(Ok(Json.toJson(madeUndertaking.reference.get)(eisCreateUndertakingResponse)))
+        Ok(Json.toJson(CreateUndertakingApiResponse(madeUndertaking.reference.get))).toFuture
     }
 
   private def getRetrieveResponse(eori: EORI) =
     eori match {
       case a if a.endsWith("999") => // fake 500
-        Future.successful(InternalServerError(Json.toJson(errorDetailFor500)))
+        InternalServerError(Json.toJson(errorDetailFor500)).toFuture
 
       case b if b.endsWith("777") => // ID invalid
-        val noUndertakingFoundResponse: JsValue = Json.obj(
-          "retrieveUndertakingResponse" -> Json.obj(
-            "responseCommon" -> badResponseCommon(
-              "055",
-              "ID number missing or invalid"
-            )
-          )
-        )
-        Future.successful(Ok(Json.toJson(noUndertakingFoundResponse)))
+        Ok(Json.toJson(RetrieveUndertakingApiResponse("055", "ID number missing or invalid"))).toFuture
+
       case c
           if c.endsWith("888") || Store.undertakings
             .retrieveByEori(eori)
             .isEmpty => // fake not found (ideally should have been 404)
-        val noUndertakingFoundResponse: JsValue = Json.obj(
-          "retrieveUndertakingResponse" -> Json.obj(
-            "responseCommon" -> badResponseCommon(
-              "107",
-              "Undertaking reference in the API not Subscribed in ETMP"
+        Ok(
+          Json.toJson(
```

Comment:
Is this nesting of `Json.toJson` intentional?

Reply:
fixed